lib/Syntax/SyntaxParsingContext.cpp

```diff
@@ -0,0 +1,201 @@
+    return Result;
+  }
+};
+
+SyntaxParsingContext::SyntaxParsingContext(bool Enabled):
+  ContextData(*new ContextInfo(Enabled)) {}
+
+SyntaxParsingContext::SyntaxParsingContext(SyntaxParsingContext &Another):
+    SyntaxParsingContext(Another.ContextData.Enabled) {}
+
+SyntaxParsingContext::~SyntaxParsingContext() { delete &ContextData; }
+
+void SyntaxParsingContext::disable() { ContextData.Enabled = false; }
+
+struct SyntaxParsingContextRoot::GlobalInfo {
+  // The source file under parsing.
+  SourceFile &File;
+
+  // All tokens in the source file. This list will shrink from the start when
+  // we start to build syntax nodes.
+  ArrayRef<RawTokenInfo> Tokens;
+
+  GlobalInfo(SourceFile &File) : File(File) {}
+  TokenSyntax retrieveTokenSyntax(SourceLoc Loc) {
+    auto TargetLoc = Loc.getOpaquePointerValue();
+    for (unsigned I = 0, N = Tokens.size(); I < N; I ++) {
+      auto Info = Tokens[I];
+      auto InfoLoc = Info.Loc.getOpaquePointerValue();
+      if (InfoLoc < TargetLoc)
+        continue;
+      assert(InfoLoc == TargetLoc);
+      Tokens = Tokens.slice(I + 1);
+      return make<TokenSyntax>(Info.Token);
+    }
+    llvm_unreachable("can not find token at Loc");
+  }
+};
+
+SyntaxParsingContextRoot::
+SyntaxParsingContextRoot(SourceFile &File, unsigned BufferID):
+    SyntaxParsingContext(File.shouldKeepTokens()),
+    GlobalData(*new GlobalInfo(File)) {
+  populateTokenSyntaxMap(File.getASTContext().LangOpts,
+                         File.getASTContext().SourceMgr,
+                         BufferID, File.AllRawTokenSyntax);
+  // Keep track of the raw tokens.
+  GlobalData.Tokens = llvm::makeArrayRef(File.AllRawTokenSyntax);
+}
+
+SyntaxParsingContextRoot::~SyntaxParsingContextRoot() {
+  std::vector<DeclSyntax> AllTopLevel;
+  if (GlobalData.File.SyntaxRoot.hasValue()) {
+    for (auto It: GlobalData.File.getSyntaxRoot().getTopLevelDecls()) {
+      AllTopLevel.push_back(It);
+    }
+  }
+  for (auto S: ContextData.PendingSyntax) {
+    std::vector<StmtSyntax> AllStmts;
+    if (S.isDecl()) {
+      AllStmts.push_back(SyntaxFactory::makeDeclarationStmt(
+        S.getAs<DeclSyntax>().getValue(), None));
+    } else if (S.isExpr()) {
+      AllStmts.push_back(SyntaxFactory::makeExpressionStmt(
+        S.getAs<ExprSyntax>().getValue(), None));
+    } else if (S.isStmt()) {
+      AllStmts.push_back(S.getAs<StmtSyntax>().getValue());
+    } else {
+      // If this is a standalone token, we create an unknown expression wrapper
+      // for it.
+      AllStmts.push_back(SyntaxFactory::makeExpressionStmt(
+        getUnknownExpr({ *S.getAs<TokenSyntax>() }), None));
+    }
+    AllTopLevel.push_back(SyntaxFactory::makeTopLevelCodeDecl(
+      SyntaxFactory::makeStmtList(AllStmts)));
+  }
+
+  Trivia Leading, Trailing;
+  GlobalData.File.SyntaxRoot.emplace(
+    SyntaxFactory::makeSourceFile(SyntaxFactory::makeDeclList(AllTopLevel),
+      SyntaxFactory::makeToken(tok::eof, "\n", SourcePresence::Present,
```

Comment:
Nit: I think it'd be better here if we made the EOF empty and give it newline trivia. The text for EOF tokens isn't serialized.

```
Trivia Leading = Trivia::newlines(1), Trailing;
GlobalData.File.SyntaxRoot.emplace(
  SyntaxFactory::makeSourceFile(SyntaxFactory::makeDeclList(AllTopLevel),
    SyntaxFactory::makeToken(tok::eof, "\n", SourcePresence::Present, Leading, Trailing)));
```